### ISSUE TITLE
Auto mounting functionality for shared folders

### ISSUE THREAD

**Issue by gbraad** [minishift/minishift#480]
1. ~~Allow configuration to auto mount shares~~
2. ~~Prevent remounting mounted shares~~
3. ~~Check if host is reachable~~
4. ~~Command to add configuration by name~~
5. ~~Command to remove configuration by name~~
6. ~~Command to mount share by name~~
7. ~~Command to unmount share by name~~
8. ~~Auto mount option for `Users` share~~

**Comment by gbraad** [comment]
Development can be tracked in the branch: https://github.com/gbraad/minishift/tree/mount-shares

  * https://github.com/gbraad/minishift/commit/cf05002d591b980c149e52224568990f475c7fae

Configuration needed:

```
{
    "iso-url": "file://C:/Users/gbraad/.minishift/cache/iso/minishift-cifs.iso",
    "vm-driver": "hyperv",
    "shares": [
        {"name": "Users", "uncpath": "//RHGBWIN10/Users", "username": "gbraad@redhat.com", "password": "********", "domain": "RHGBWIN10"},
        {"name": "Disks", "uncpath": "//10.0.21.33/disks", "username": "root", "password": "********", "domain": ""}
    ]
}
```

**Comment by gbraad** [comment]
@hferentschik @LalatenduMohanty I would not really like to introduce a new verb as command, but I think this is needed? ...

```
$ minishift mount [params]
$ minishift unmount [name]
```
This feels weird...

So, what about?
```
$ minishift share mount [name]
$ minishift share unmount [name]
```

Or place them inside the `config` command?

And how to structure them for edit and save? WDYT?

**Comment by hferentschik** [comment]
>  I would not really like to introduce a new verb as command

Sure. How do you want to use that? Once Minishift is up you do mount/umount? I guess I am trying to figure out how this is tied into the current behavior of say VirtualBox and that it transparently mounts your home directory (same for xhyve with the appropriate flags). 

Will host folder mounts on Windows with Samba shares work differently (in terms off when and how they are established) than with VirtualBox and guest additions? Or moving forward with sshfs?

@gbraad maybe you can outline how current host folder mount behavior is going to be unified with what you are doing now and this new command. 

Generally, I like the idea. I like the idea of having the ability to potentially umount a directory and mount a different one.  What about something like

    $ minishift hostfolder [view|mount|umount]

So I am proposing _hostfolder_ as command name which seems more generic. 'share' might be pointing into Windows share and afaiu this is more generic. _view_ let's you view the current host folders and _mount_ and _umount_ do the mounting and unmounting. 

_mount_ would also have a flag _type_ which lets you select between say windows share, sshfs, nfs (xhyve).

> Or place them inside the config command?

It's a possibility, but I don't think it works with the existing _set_. You would still need _config mount_, right? Seems not so intuitive to me.

**Comment by hferentschik** [comment]
Another question in my mind. How is Windows shares host folder mounting working together with another potentially exiting host folder mount. Taking VirtualBox again. If I start the default B2D image on Windows, what's going to happen?

**Comment by gbraad** [comment]
Implemented in https://github.com/gbraad/minishift/commit/bbec16711170ee64478e3ec8ef5b1236f02cfa30:

* `minishift hostfolder`
  * `mount name`
  * `umount name`

```powershell
PS > C:\work\bin\minishift.exe ssh "mount | grep Disks"
PS > C:\work\bin\minishift.exe hostfolder mount Disks
-- Mounting 'Disks': '//10.0.21.33/disks' ... OK
PS > C:\work\bin\minishift.exe ssh "mount | grep Disks"
//10.0.21.33/disks on /mnt/sda1/Disks type cifs (...)
PS > C:\work\bin\minishift.exe hostfolder umount Disks
-- Unmounting 'Disks': '//10.0.21.33/disks' ... OK
PS > C:\work\bin\minishift.exe ssh "mount | grep Disks"
PS >
```

**Comment by gbraad** [comment]
Will finish the mount check and IP reachability, and then defer configuration to instance specific helper functionality (see #487)

**Comment by gbraad** [comment]
```
$ minishift hostfolder mount unreachable
Mounting 'unreachable': '//unreachable/disks' ... Unreachable
$ minishift hostfolder umount unreachable
Unmounting 'unreachable': '//unreachable/disks' ... Not mounted
```

```
$ minishift hostfolder mount Disks
Mounting 'Disks': '//10.0.21.33/disks' ... Already mounted
$ minishift hostfolder umount Disks
Unmounting 'Disks': '//10.0.21.33/disks' ... OK
$ minishift hostfolder mount Disks
Mounting 'Disks': '//10.0.21.33/disks' ... OK
$ minishift hostfolder mount Disks
Mounting 'Disks': '//10.0.21.33/disks' ... Already mounted
```

**Comment by gbraad** [comment]
Currently remaining is: *Auto mount option for Users share*
I have an idea how to tackle this:

  * if driver is hyper-v
    * check if not Users already configed
    * get hostips (check reachable)
    * get hostip in range of instance ip
    * construct uncpath
    * ask for credentials
    * store as preserved hostfolder name: Users

**Comment by hferentschik** [comment]
> if driver is hyper-v

this works only for Hyper-V?

**Comment by gbraad** [comment]
> only for Hyper-V
@hferentschik no, I'd rather not... but not sure about conflicts when VirtualBox Guest Additions is used.

Will test with Virtualbox today... Just pushed a WIP for feedback which used the instance configuration. Will move to VirtualBox for some testing.

**Comment by hferentschik** [comment]
> Will move to VirtualBox for some testing. None of the current code enforces the libmachine Driver.

right, that's what I was assuming and hoping for ;-)

**Comment by gbraad** [comment]
test command for `hostip` now only returns a host IP address in the same network as the minishift instance. Eg.

```powershell
PS C:\work\src\github.com\minishift\minishift> minishift.exe hostip
   Checking if '10.0.21.122' is reachable ... OK
```

this will be used for the construction of the uncpath to the local shares for `//hostip/Users`, but likely also for @hferentschik's sshfs path ;-).

**Comment by budhrg** [comment]
@gbraad  :+1:  Nice

I would love to have this output though :wink: 
```
..> minishift.exe hostip
10.0.21.122
```

**Comment by gbraad** [comment]
@budhrg this command will disappear...

**Comment by gbraad** [comment]
Added config add/remove functionality

**Comment by gbraad** [comment]
Added `Users` logic which treats the name as a special case and takes the IP address as part of the UNC path.

Code really needs to be refactored and hardened for some error cases, but generally it works...

**Comment by gbraad** [comment]
@budhrg 

On Linux:
```
 $ minishift ip
192.168.42.69
$ minishift hostip
192.168.42.1
```

On Windows
```
PS C:\work\src\github.com\minishift\minishift> minishift ip
10.0.21.123
PS C:\work\src\github.com\minishift\minishift> minishift hostip
10.0.21.124
```

```
PS C:\work\src\github.com\minishift\minishift> minishift hostfolder add Users
UNC path:  //10.0.21.124/Users
Username: gbraad@redhat.com
Password: 
Domain: RHGBWIN10
Added: Users
PS C:\work\src\github.com\minishift\minishift> minishift hostfolder view
Name            Mountpoint              UNC path
Users   /mnt/sda1/Users         //10.0.21.124/Users
```

Note: found a small issue on Windows with the crypto terminal library.

**Comment by gbraad** [comment]
`x/crypto/ssh/terminal: ReadPassword broken on Windows`
  * https://github.com/golang/go/issues/11914
  * https://github.com/golang/go/issues/16552

**Comment by gbraad** [comment]
I have refactored some of the code to be more generic and a clearer distinction between implementation specific code. However, I will still have several other passes of optimization and refactoring as still duplicate/redundant code is present.

**Comment by gbraad** [comment]
Running state is enforced, error states returned, `Users` is only determined when running (record is used to store credentials), added a `mount-all` command (eg. for when auto-mount is disabled):

```
minishift hostfolder mount-all
-- Mounting hostfolders
   Mounting 'disks': '//10.0.21.33/Disks' ... Already mounted
   Mounting 'Users': '//192.168.42.1/Users' ... FAIL
   Mounting 'unreachable': '//10.0.21.34/unreachable' ... Unreachable
```

**Comment by hferentschik** [comment]
Merged via pull request #581. 👍

**Comment by Jhstvf4t** [comment]
sorry boss, I ran out of money in and out of the workshop, what would I eat if the money was only for repairing motorbikes etc
